File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/picker_dialog_adapter/AdapterPickerDialogBuilder.java
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_adapter;


import com.github.rooneyandshadows.lightbulb.dialogs.base.LightBulbDialogBuilder;
import com.github.rooneyandshadows.lightbulb.dialogs.base.LightBulbDialogFragment;
import com.github.rooneyandshadows.lightbulb.recycleradapters.LightBulbAdapter;
import com.github.rooneyandshadows.lightbulb.recycleradapters.LightBulbAdapterDataModel;

import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.RecyclerView;

public class AdapterPickerDialogBuilder<ModelType extends LightBulbAdapterDataModel> extends LightBulbDialogBuilder<AdapterPickerDialog<ModelType>> {
    private final LightBulbAdapter<ModelType> adapter;
    private AdapterPickerDialog.SelectionChangedListener<int[]> changedCallback;
    private RecyclerView.ItemDecoration itemDecoration;
    private int[] selection;

    public AdapterPickerDialogBuilder(FragmentManager manager, String dialogTag, LightBulbAdapter<ModelType> adapter) {
        super(manager, dialogTag);
        this.adapter = adapter;
    }

    @Override
    public AdapterPickerDialogBuilder<ModelType> withTitle(String title) {
        return (AdapterPickerDialogBuilder<ModelType>) super.withTitle(title);
    }

    @Override
    public AdapterPickerDialogBuilder<ModelType> withMessage(String message) {
        return (AdapterPickerDialogBuilder<ModelType>) super.withMessage(message);
    }

    @Override
    public AdapterPickerDialogBuilder<ModelType> withPositiveButton(LightBulbDialogFragment.DialogButtonConfiguration positiveButtonConfiguration, LightBulbDialogFragment.DialogButtonClickListener onClickListener) {
        return (AdapterPickerDialogBuilder<ModelType>) super.withPositiveButton(positiveButtonConfiguration, onClickListener);
    }

    @Override
    public AdapterPickerDialogBuilder<ModelType> withNegativeButton(LightBulbDialogFragment.DialogButtonConfiguration negativeButtonConfiguration, LightBulbDialogFragment.DialogButtonClickListener onClickListener) {
        return (AdapterPickerDialogBuilder<ModelType>) super.withNegativeButton(negativeButtonConfiguration, onClickListener);
    }

    @Override
    public AdapterPickerDialogBuilder<ModelType> withOnCancelListener(LightBulbDialogFragment.DialogCancelListener listener) {
        return (AdapterPickerDialogBuilder<ModelType>) super.withOnCancelListener(listener);
    }

    @Override
    public AdapterPickerDialogBuilder<ModelType> withOnShowListener(LightBulbDialogFragment.DialogShowListener listener) {
        return (AdapterPickerDialogBuilder<ModelType>) super.withOnShowListener(listener);
    }

    @Override
    public AdapterPickerDialogBuilder<ModelType> withOnHideListener(LightBulbDialogFragment.DialogHideListener listener) {
        return (AdapterPickerDialogBuilder<ModelType>) super.withOnHideListener(listener);
    }

    @Override
    public AdapterPickerDialogBuilder<ModelType> withCancelOnClickOutsude(boolean closeOnClickOutside) {
        return (AdapterPickerDialogBuilder<ModelType>) super.withCancelOnClickOutsude(closeOnClickOutside);
    }

    @Override
    public AdapterPickerDialogBuilder<ModelType> withDialogType(LightBulbDialogFragment.DialogTypes dialogType) {
        return (AdapterPickerDialogBuilder<ModelType>) super.withDialogType(dialogType);
    }

    @Override
    public AdapterPickerDialogBuilder<ModelType> withAnimations(LightBulbDialogFragment.DialogAnimationTypes animation) {
        return (AdapterPickerDialogBuilder<ModelType>) super.withAnimations(animation);
    }

    public AdapterPickerDialogBuilder<ModelType> withSelectionCallback(AdapterPickerDialog.SelectionChangedListener<int[]> listener) {
        changedCallback = listener;
        return this;
    }

    public AdapterPickerDialogBuilder<ModelType> withSelection(int[] selection) {
        this.selection = selection;
        return this;
    }

    public AdapterPickerDialogBuilder<ModelType> withItemDecoration(RecyclerView.ItemDecoration decoration) {
        this.itemDecoration = decoration;
        return this;
    }


    @Override
    @SuppressWarnings("unchecked")
    public AdapterPickerDialog<ModelType> buildDialog() {
        AdapterPickerDialog<ModelType> dialogFragment = (AdapterPickerDialog<ModelType>) fragmentManager.findFragmentByTag(dialogTag);
        if (dialogFragment == null)
            dialogFragment = AdapterPickerDialog.newInstance(title, message, positiveButtonConfiguration, negativeButtonConfiguration, cancelableOnClickOutside, dialogType, animation);
        dialogFragment.setFragmentManager(fragmentManager);
        dialogFragment.setDialogTag(dialogTag);
        dialogFragment.addOnNegativeClickListeners(onNegativeClickListener);
        dialogFragment.addOnPositiveClickListener(onPositiveClickListener);
        dialogFragment.addOnShowListener(onShowListener);
        dialogFragment.addOnHideListener(onHideListener);
        dialogFragment.addOnCancelListener(onCancelListener);
        dialogFragment.setAdapter(adapter);
        dialogFragment.setOnSelectionChangedListener(changedCallback);
        dialogFragment.setSelection(selection);
        dialogFragment.setItemDecoration(itemDecoration);
        return dialogFragment;
    }
}

File: lightbulb-dialogs/src/main/java/com/github/rooneyandshadows/lightbulb/dialogs/picker_dialog_adapter/AdapterPickerDialogBuilder.kt
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_adapter

import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.LifecycleOwner
import androidx.recyclerview.widget.RecyclerView
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment.SelectionChangedListener
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.*
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.callbacks.*
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.data.EasyAdapterDataModel

@Suppress("unused")
class AdapterPickerDialogBuilder<DialogType : AdapterPickerDialog<out EasyAdapterDataModel>> @JvmOverloads constructor(
    dialogTag: String,
    dialogParentFragmentManager: FragmentManager,
    private val dialogInitializer: AdapterPickerDialogInitializer<DialogType>,
    dialogLifecycleOwner: LifecycleOwner? = null,
    initialDialogState: Bundle? = null
) : BaseDialogBuilder<DialogType>(dialogTag, dialogParentFragmentManager, dialogLifecycleOwner, initialDialogState) {
    private var changedCallback: SelectionChangedListener<IntArray>? = null
    private var itemDecoration: RecyclerView.ItemDecoration? = null
    private var selection: IntArray? = null

    @JvmOverloads
    constructor(
        dialogTag: String,
        fragment: Fragment,
        dialogInitializer: AdapterPickerDialogInitializer<DialogType>,
        initialDialogState: Bundle? = null
    ) : this(
        dialogTag,
        fragment.childFragmentManager,
        dialogInitializer,
        fragment,
        initialDialogState
    )

    @JvmOverloads
    constructor(
        dialogTag: String,
        activity: FragmentActivity,
        dialogInitializer: AdapterPickerDialogInitializer<DialogType>,
        initialDialogState: Bundle? = null
    ) : this(
        dialogTag,
        activity.supportFragmentManager,
        dialogInitializer,
        activity,
        initialDialogState
    )

    @Override
    override fun setupNonRetainableSettings(dialog: DialogType) {
        dialog.apply {
            changedCallback?.apply { addOnSelectionChangedListener(this) }
            setItemDecoration(itemDecoration)
        }
    }

    @Override
    override fun setupRetainableSettings(dialog: DialogType) {
        dialog.apply {
            setSelection(selection)
        }
    }

    @Override
    override fun initializeNewDialog(): DialogType {
        return dialogInitializer.initialize()
    }

    @Override
    override fun withTitle(title: String): AdapterPickerDialogBuilder<DialogType> {
        return super.withTitle(title) as AdapterPickerDialogBuilder<DialogType>
    }

    @Override
    override fun withMessage(message: String): AdapterPickerDialogBuilder<DialogType> {
        return super.withMessage(message) as AdapterPickerDialogBuilder<DialogType>
    }

    @Override
    override fun withButton(configuration: DialogButton): AdapterPickerDialogBuilder<DialogType> {
        return super.withButton(configuration) as AdapterPickerDialogBuilder<DialogType>
    }

    @Override
    override fun withOnCancelListener(listener: DialogCancelListener): AdapterPickerDialogBuilder<DialogType> {
        return super.withOnCancelListener(listener) as AdapterPickerDialogBuilder<DialogType>
    }

    @Override
    override fun withOnShowListener(listener: DialogShowListener): AdapterPickerDialogBuilder<DialogType> {
        return super.withOnShowListener(listener) as AdapterPickerDialogBuilder<DialogType>
    }

    @Override
    override fun withOnHideListener(listener: DialogHideListener): AdapterPickerDialogBuilder<DialogType> {
        return super.withOnHideListener(listener) as AdapterPickerDialogBuilder<DialogType>
    }

    @Override
    override fun withCancelOnClickOutside(closeOnClickOutside: Boolean): AdapterPickerDialogBuilder<DialogType> {
        return super.withCancelOnClickOutside(closeOnClickOutside) as AdapterPickerDialogBuilder<DialogType>
    }

    @Override
    override fun withDialogType(dialogType: DialogTypes): AdapterPickerDialogBuilder<DialogType> {
        return super.withDialogType(dialogType) as AdapterPickerDialogBuilder<DialogType>
    }

    @Override
    override fun withAnimations(animation: DialogAnimationTypes): AdapterPickerDialogBuilder<DialogType> {
        return super.withAnimations(animation) as AdapterPickerDialogBuilder<DialogType>
    }

    @Override
    override fun withDialogListeners(listeners: DialogListeners): AdapterPickerDialogBuilder<DialogType> {
        return super.withDialogListeners(listeners) as AdapterPickerDialogBuilder<DialogType>
    }

    fun withSelectionCallback(listener: SelectionChangedListener<IntArray>): AdapterPickerDialogBuilder<DialogType> {
        changedCallback = listener
        return this
    }

    fun withSelection(selection: IntArray): AdapterPickerDialogBuilder<DialogType> {
        this.selection = selection
        return this
    }

    fun withItemDecoration(decoration: RecyclerView.ItemDecoration?): AdapterPickerDialogBuilder<DialogType> {
        itemDecoration = decoration
        return this
    }

    fun interface AdapterPickerDialogInitializer<DialogType : AdapterPickerDialog<out EasyAdapterDataModel>> {
        fun initialize(): DialogType
    }
}